java/statestore/src/main/java/sleeper/statestore/transactionlog/TransactionLogSnapshotCreator.java

```diff
@@ -0,0 +1,136 @@
+
+public class TransactionLogSnapshotCreator {
+    public static final Logger LOGGER = LoggerFactory.getLogger(TransactionLogSnapshotCreator.class);
+    private final InstanceProperties instanceProperties;
+    private final TableProperties tableProperties;
+    private final TransactionLogStore filesLogStore;
+    private final TransactionLogStore partitionsLogStore;
+    private final TransactionLogSnapshotSerDe snapshotSerDe;
+    private final DynamoDBTransactionLogSnapshotStore snapshotStore;
+
+    public TransactionLogSnapshotCreator(
+            InstanceProperties instanceProperties, TableProperties tableProperties,
+            AmazonS3 s3Client, AmazonDynamoDB dynamoDB, Configuration configuration) {
+        this.instanceProperties = instanceProperties;
+        this.tableProperties = tableProperties;
+        this.filesLogStore = new DynamoDBTransactionLogStore(instanceProperties.get(TRANSACTION_LOG_FILES_TABLENAME),
+                instanceProperties, tableProperties, dynamoDB, s3Client);
+        this.partitionsLogStore = new DynamoDBTransactionLogStore(instanceProperties.get(TRANSACTION_LOG_PARTITIONS_TABLENAME),
+                instanceProperties, tableProperties, dynamoDB, s3Client);
+        this.snapshotSerDe = new TransactionLogSnapshotSerDe(tableProperties.getSchema(), configuration);
+        this.snapshotStore = new DynamoDBTransactionLogSnapshotStore(instanceProperties, tableProperties, dynamoDB);
+    }
+
+    public void createSnapshot() {
+        LOGGER.info("Creating snapshot for table {}", tableProperties.getStatus());
+        Optional<LatestSnapshots> latestSnapshotsOpt = snapshotStore.getLatestSnapshots();
+        StateStoreFiles filesState = latestSnapshotsOpt
+                .filter(latestSnapshot -> latestSnapshot.getFilesSnapshot() != null)
+                .map(latestSnapshot -> {
+                    try {
+                        return snapshotSerDe.loadFiles(latestSnapshot.getFilesSnapshot());
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                })
+                .orElseGet(StateStoreFiles::new);
+        long filesTransactionNumberBefore = latestSnapshotsOpt
+                .filter(latestSnapshot -> latestSnapshot.getFilesSnapshot() != null)
+                .map(latestSnapshot -> latestSnapshot.getFilesSnapshot().getTransactionNumber())
+                .orElse(0L);
+
+        StateStorePartitions partitionsState = latestSnapshotsOpt
+                .filter(latestSnapshot -> latestSnapshot.getPartitionsSnapshot() != null)
+                .map(latestSnapshot -> {
+                    try {
+                        return snapshotSerDe.loadPartitions(latestSnapshot.getPartitionsSnapshot());
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                })
+                .orElseGet(StateStorePartitions::new);
```

Comment:
This logic could be clearer if getLatestSnapshots returned an object that can be empty itself, rather than an optional.